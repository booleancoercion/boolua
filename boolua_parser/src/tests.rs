use insta::assert_debug_snapshot;
use paste::paste;

macro_rules! parse_and_verify {
    ($name:ident, $path:literal) => {
        paste! {
            #[test]
            fn [<parse_and_verify_ $name>]() {
                let source = std::fs::read_to_string($path).unwrap();
                let ast = crate::parse_source(&source).expect("parsing generated errors");
                assert_debug_snapshot!(ast);
            }
        }
    };
}

parse_and_verify!(simple_if, "../programs/simple_if.lua");
parse_and_verify!(simple_expr, "../programs/simple_expr.lua");
parse_and_verify!(numbers, "../programs/numbers.lua");
parse_and_verify!(ambiguity1, "../programs/ambiguous.lua");
parse_and_verify!(ambiguity2, "../programs/unambiguous.lua");
