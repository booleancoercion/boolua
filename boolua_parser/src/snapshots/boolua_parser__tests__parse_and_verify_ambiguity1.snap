---
source: src/tests.rs
assertion_line: 20
expression: ast

---
Block {
    stmts: [
        Assignment(
            [
                Simple(
                    Name("a"),
                ),
            ],
            [
                Binary(
                    Prefix(
                        Var(
                            Simple(
                                Name("b"),
                            ),
                        ),
                    ),
                    Add,
                    Prefix(
                        FnCall(
                            FnCall(
                                FnCall(
                                    FnCall(
                                        Var(
                                            Simple(
                                                Name("c"),
                                            ),
                                        ),
                                        None,
                                        [
                                            Binary(
                                                Prefix(
                                                    Var(
                                                        Simple(
                                                            Name("print"),
                                                        ),
                                                    ),
                                                ),
                                                Or,
                                                Prefix(
                                                    Var(
                                                        Indexed(
                                                            Var(
                                                                Simple(
                                                                    Name("io"),
                                                                ),
                                                            ),
                                                            Str(
                                                                "write",
                                                            ),
                                                        ),
                                                    ),
                                                ),
                                            ),
                                        ],
                                    ),
                                ),
                                None,
                                [
                                    Str(
                                        "done",
                                    ),
                                ],
                            ),
                        ),
                    ),
                ),
            ],
        ),
    ],
    ret: None,
}
